patroni/config_generator.py

```diff
@@ -0,0 +1,290 @@
+              Sorting guarantees it will prefer ipv4.
+    """
+    hostname = None
+    try:
+        hostname = socket.gethostname()
+        return sorted(socket.getaddrinfo(hostname, 0, socket.AF_UNSPEC, socket.SOCK_STREAM, 0),
+                      key=lambda x: x[0])[0][4][0]
+    except OSError as e:
+        sys.exit(f'Failed to define ip address: {e}')
+    except IndexError:
+        sys.exit(f'Failed to define ip address. No address returned by getaddrinfo for {hostname}')
+
+
+def get_int_major_version(config: Optional[Dict[str, Any]] = None) -> int:
+    """Get major version of PostgreSQL from the binary as an integer.
+
+    :param config: optional dictionary representing Patroni config. If contains bin_dir and/or the custom
+    bin_name for postgres, the values are used for the version retrieval.
+    """
+    config = config or {}
+    postgres_bin = ((config.get('postgresql') or {}).get('bin_name') or {}).get('postgres', 'postgres')
+    return postgres_major_version_to_int(get_major_version(config['postgresql'].get('bin_dir') or None, postgres_bin))
+
+
+def get_bin_dir_from_running_instance(data_dir: str) -> str:
+    """Define the directory postgres binaries reside using postmaster's pid executable."""
+    postmaster_pid = None
+    try:
+        with open(f"{data_dir}/postmaster.pid", 'r') as f:
+            postmaster_pid = f.readline()
+            if not postmaster_pid:
+                sys.exit('Failed to obtain postmaster pid from postmaster.pid file')
+            postmaster_pid = int(postmaster_pid.strip())
+    except OSError as e:
+        sys.exit(f'Error while reading postmaster.pid file: {e}')
+    import psutil
+    try:
+        return os.path.dirname(psutil.Process(postmaster_pid).exe())
+    except psutil.NoSuchProcess:
+        sys.exit('Obtained postmaster pid doesn\'t exist')
+
+
+def enrich_config_from_running_instance(config: Dict[str, Any], no_value_msg: str, dsn: Optional[str] = None) -> None:
+    """Extend the passed *config* dictionary with the values gathered from a running instance.
+
+    Get:
+
+    - non-internal GUC values having configuration file, postmaster command line or environment variable as a source
+    - ``postgresql.connect_address``, postgresql.listen``,
+    - ``postgresql.pg_hba`` and ``postgresql.pg_ident`` if hba_file/ident_file is set to the default value
+    - superuser auth parameters (from the options used for connection)
+
+    And redefine scope with the ``cluster_name`` GUC value if set.
+
+    :param config: configuration parameters dict to be enriched
+    :param no_value_msg: str value to be used when a parameter value is not available
+    :param dsn: optional DSN string for the source running instance
+    """
+    from getpass import getuser, getpass
+    from patroni.postgresql.config import parse_dsn
+    from patroni.config import AUTH_ALLOWED_PARAMETERS_MAPPING
```

Comment:
I wonder if we should move this to the top.

Reply:
>  It is strange to me that all imports are done in methods in Patroni.

In Patroni it is usually done not to import when it is actually not required (some branches are not that often accessed)
As I understand it (at least that is what guides me to do it like that) @CyberDem0n?

Reply:
I wonder how efficient that is. If you need to run import machinery to load a module at time of execution vs once when the application starts it might actually add a time hit. An imported module is a single instance in memory so once it's imported the same module is accessed on each import, rather than read from disk every import statement. That could potentially still cause a performance hit in some cases, although negligible. Obviously, there's different visible effects between the user cli, api and the daemon. Just out of interest, has any benchmarking/profiling been done for this?

Reply:
The problem exists only with `patroni/__main__.py`, which could fork Patroni when running with the PID=1 (in docker).

That is, https://github.com/zalando/patroni/pull/2786/files#diff-48810aca58351596dee0cb6815e887053509d54b64929c9f1fd8a2c1c538a752R10 should be imported from process_arguments(). Inside the config_generator we can do anything you like.

Reply:
🤦‍♀️ how short memory I have